File: codebase/src/cpp/hla13/test/ddm/DDMSupportServicesTest.cpp
/*
 *   Copyright 2008 The Portico Project
 *
 *   This file is part of portico.
 *
 *   portico is free software; you can redistribute it and/or modify
 *   it under the terms of the Common Developer and Distribution License (CDDL)
 *   as published by Sun Microsystems. For more information see the LICENSE file.
 *
 *   Use of this software is strictly AT YOUR OWN RISK!!!
 *   If something bad happens you do not have permission to come crying to me.
 *   (that goes for your lawyer as well)
 *
 */
#include "DDMSupportServicesTest.h"

// Register test suite with the global repository
CPPUNIT_TEST_SUITE_REGISTRATION( DDMSupportServicesTest );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( DDMSupportServicesTest, "DDMSupportServicesTest" );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( DDMSupportServicesTest, "ddm" );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( DDMSupportServicesTest, "supportServices" );

/////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////// Constructors/Destructors /////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////
DDMSupportServicesTest::DDMSupportServicesTest()
{
	this->defaultFederate = new Test13Federate( "defaultFederate" );
}

DDMSupportServicesTest::~DDMSupportServicesTest()
{
	delete this->defaultFederate;
}

/////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////// Test Setup and Helper Methods ///////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////

void DDMSupportServicesTest::setUp()
{
	this->defaultFederate->quickCreate();
	this->defaultFederate->quickJoin();
}

void DDMSupportServicesTest::tearDown()
{
	this->defaultFederate->quickResign();
	this->defaultFederate->quickDestroy();
}

////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////// Test Methods ///////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////

///////////////////////////////////
// TEST: (valid) testGetRegion() //
///////////////////////////////////
// RTI::Region* getRegion( RTI::RegionToken token )
//     throw( RTI::FederateNotExecutionMember,
//            RTI::ConcurrentAccessAttempted,
//            RTI::RegionNotKnown,
//            RTI::RTIinternalError )
void DDMSupportServicesTest::testGetRegion()
{
	// register a region
	RTI::Region *region = defaultFederate->quickCreateTestRegion( 100, 150 );
	RTI::RegionToken token = defaultFederate->quickGetRegionToken( region );
	RTI::Region *otherRegion = NULL; 
	
	// get the region for the token and make sure they are the same
	try
	{
		otherRegion = defaultFederate->rtiamb->getRegion( token );
	}
	catch( RTI::Exception &e )
	{
		delete region;
		failTest( "Fetching region for valid token: %s", e._reason );
	}
	
	if( otherRegion == NULL )
	{
		delete region;
		failTest( "Failed to fetch region, null found" );
	}
	
	// make sure the values are the same
	CPPUNIT_ASSERT_EQUAL( region->getSpaceHandle(), otherRegion->getSpaceHandle() );
	CPPUNIT_ASSERT_EQUAL( region->getNumberOfExtents(), otherRegion->getNumberOfExtents() );
	
	delete region;
	delete otherRegion;
}

////////////////////////////////////////
// TEST: (valid) testGetRegionToken() //
////////////////////////////////////////
// RTI::RegionToken getRegionToken( RTI::Region *theRegion )
//     throw( RTI::FederateNotExecutionMember, 
//            RTI::ConcurrentAccessAttempted,
//            RTI::RegionNotKnown,
//            RTI::RTIinternalError )
void DDMSupportServicesTest::testGetRegionToken()
{
	RTI::Region *region = defaultFederate->quickCreateTestRegion( 100, 150 );
	RTI::RegionToken token = defaultFederate->rtiamb->getRegionToken( region );
	RTI::Region *other = defaultFederate->rtiamb->getRegion( token );
	
	CPPUNIT_ASSERT_EQUAL( region->getNumberOfExtents(), other->getNumberOfExtents() );
	CPPUNIT_ASSERT_EQUAL( region->getSpaceHandle(), other->getSpaceHandle() );
	
	delete other;
	delete region;
}

////////////////////////////////////////////////////////
// TEST: (valid) testGetAttributeRoutingSpaceHandle() //
////////////////////////////////////////////////////////
// RTI::SpaceHandle
// getAttributeRoutingSpaceHandle( RTI::AttributeHandle theHandle, 
//                                 RTI::ObjectClassHandle whichClass )
//     throw( RTI::ObjectClassNotDefined,
//            RTI::AttributeNotDefined,
//            RTI::FederateNotExecutionMember,
//            RTI::ConcurrentAccessAttempted,
//            RTI::RTIinternalError )
void DDMSupportServicesTest::testGetAttributeRoutingSpaceHandle()
{
	RTI::SpaceHandle expected = defaultFederate->quickSpaceHandle( "TestSpace" );
	RTI::ObjectClassHandle classHandle = defaultFederate->quickOCHandle( "ObjectRoot.A" );
	RTI::AttributeHandle attributeHandle = defaultFederate->quickACHandle( "ObjectRoot.A", "aa" );
	
	RTI::SpaceHandle actual =
		defaultFederate->rtiamb->getAttributeRoutingSpaceHandle( attributeHandle, classHandle );
	
	CPPUNIT_ASSERT_EQUAL( expected, actual );
}

////////////////////////////////////////////
// TEST: (valid) testGetDimensionHandle() //
////////////////////////////////////////////
// RTI::DimensionHandle
// getDimensionHandle( const char *theName, RTI::SpaceHandle whichSpace )
//     throw( RTI::SpaceNotDefined,
//            RTI::NameNotFound,
//            RTI::FederateNotExecutionMember,
//            RTI::ConcurrentAccessAttempted,
//            RTI::RTIinternalError )
void DDMSupportServicesTest::testGetDimensionHandle()
{
	RTI::SpaceHandle spaceHandle = defaultFederate->quickSpaceHandle( "TestSpace" );
	RTI::DimensionHandle dimensionHandle =
		defaultFederate->rtiamb->getDimensionHandle( "TestDimension", spaceHandle );
	
	CPPUNIT_ASSERT_EQUAL( (RTI::ULong)(spaceHandle+1), (RTI::ULong)dimensionHandle );
}

//////////////////////////////////////////
// TEST: (valid) testGetDimensionName() //
//////////////////////////////////////////
// char*
// getDimensionName( RTI::DimensionHandle theHandle, RTI::SpaceHandle whichSpace )
//     throw( RTI::SpaceNotDefined,
//            RTI::DimensionNotDefined,
//            RTI::FederateNotExecutionMember,
//            RTI::ConcurrentAccessAttempted,
//            RTI::RTIinternalError )
void DDMSupportServicesTest::testGetDimensionName()
{
	RTI::SpaceHandle spaceHandle = defaultFederate->quickSpaceHandle( "TestSpace" );
	RTI::DimensionHandle dimensionHandle =
		defaultFederate->quickDimensionHandle( "TestSpace", "TestDimension" );
	
	char *dimensionName = defaultFederate->rtiamb->getDimensionName( dimensionHandle, spaceHandle );
	int result = strcmp( "TestDimension", dimensionName );
	
	CPPUNIT_ASSERT_EQUAL( 0, result );
}

//////////////////////////////////////////////////////////
// TEST: (valid) testGetInteractionRoutingSpaceHandle() //
//////////////////////////////////////////////////////////
// RTI::SpaceHandle
// getInteractionRoutingSpaceHandle( RTI::InteractionClassHandle theHandle )
//     throw( RTI::InteractionClassNotDefined,
//            RTI::FederateNotExecutionMember,
//            RTI::ConcurrentAccessAttempted,
//            RTI::RTIinternalError )
void DDMSupportServicesTest::testGetInteractionRoutingSpaceHandle()
{
	RTI::SpaceHandle expected = defaultFederate->quickSpaceHandle( "TestSpace" );
	RTI::ObjectClassHandle classHandle = defaultFederate->quickICHandle( "InteractionRoot.X" );

	RTI::SpaceHandle actual = defaultFederate->rtiamb->getInteractionRoutingSpaceHandle( classHandle );

	CPPUNIT_ASSERT_EQUAL( expected, actual );
}

///////////////////////////////////////////////
// TEST: (valid) testGetRoutingSpaceHandle() //
///////////////////////////////////////////////
// RTI::SpaceHandle getRoutingSpaceHandle( const char *theName )
//     throw( RTI::NameNotFound,
//            RTI::FederateNotExecutionMember,
//            RTI::ConcurrentAccessAttempted,
//            RTI::RTIinternalError )
void DDMSupportServicesTest::testGetRoutingSpaceHandle()
{
	// not really sure how to test this. just same as getSpaceName() until I figure something out
	RTI::SpaceHandle spaceHandle = defaultFederate->quickSpaceHandle( "TestSpace" );
	char *spaceName = defaultFederate->rtiamb->getRoutingSpaceName( spaceHandle );
	int result = strcmp( "TestSpace", spaceName );
	
	CPPUNIT_ASSERT_EQUAL( 0, result );
}

/////////////////////////////////////////////
// TEST: (valid) testGetRoutingSpaceName() //
/////////////////////////////////////////////
// char* getRoutingSpaceName( RTI::SpaceHandle theHandle )
//     throw( RTI::SpaceNotDefined,
//            RTI::FederateNotExecutionMember,
//            RTI::ConcurrentAccessAttempted,
//            RTI::RTIinternalError )
void DDMSupportServicesTest::testGetRoutingSpaceName()
{
	RTI::SpaceHandle spaceHandle = defaultFederate->quickSpaceHandle( "TestSpace" );
	
	char *spaceName = defaultFederate->rtiamb->getRoutingSpaceName( spaceHandle );
	int result = strcmp( "TestSpace", spaceName );
	
	CPPUNIT_ASSERT_EQUAL( 0, result );
}

File: codebase/src/java/test/hlaunit/hla13/ddm/DDMSupportServicesTest.java
/*
 *   Copyright 2007 The Portico Project
 *
 *   This file is part of portico.
 *
 *   portico is free software; you can redistribute it and/or modify
 *   it under the terms of the Common Developer and Distribution License (CDDL) 
 *   as published by Sun Microsystems. For more information see the LICENSE file.
 *   
 *   Use of this software is strictly AT YOUR OWN RISK!!!
 *   If something bad happens you do not have permission to come crying to me.
 *   (that goes for your lawyer as well)
 *
 */
package hlaunit.hla13.ddm;

import hla.rti.DimensionNotDefined;
import hla.rti.FederateNotExecutionMember;
import hla.rti.NameNotFound;
import hla.rti.Region;
import hla.rti.RegionNotKnown;
import hla.rti.SpaceNotDefined;
import hlaunit.hla13.common.Abstract13Test;
import hlaunit.hla13.common.Test13Federate;

import org.testng.Assert;
import org.testng.annotations.AfterClass;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

/**
 * These tests are used to validate the DDM-related support service methods in the RTIambassador
 * API.
 */
@Test(singleThreaded=true,groups={"DDMSupportServicesTest", "ddm","supportServices"})
public class DDMSupportServicesTest extends Abstract13Test
{
	//----------------------------------------------------------
	//                    STATIC VARIABLES
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                   INSTANCE VARIABLES
	//----------------------------------------------------------
	private Test13Federate secondFederate;
	private int spaceHandle;
	private int dimensionHandle;
	private Region region;
	private int regionHandle;

	//----------------------------------------------------------
	//                      CONSTRUCTORS
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                    INSTANCE METHODS
	//----------------------------------------------------------
	@BeforeClass(alwaysRun=true)
	public void beforeClass()
	{
		super.beforeClass();
		
		// initialize the second federate
		this.secondFederate = new Test13Federate( "secondFederate", this );
	}
	
	@BeforeMethod(alwaysRun=true)
	public void beforeMethod()
	{
		defaultFederate.quickCreate();
		defaultFederate.quickJoin();
		secondFederate.quickJoin();
		
		// create the region
		spaceHandle = defaultFederate.quickSpaceHandle( "TestSpace" );
		dimensionHandle = defaultFederate.quickDimensionHandle( "TestSpace", "TestDimension" );
		region = defaultFederate.quickCreateRegion( spaceHandle, 1 );
		regionHandle = defaultFederate.quickGetRegionToken( region );
	}
	
	@AfterMethod(alwaysRun=true)
	public void afterMethod()
	{
		secondFederate.quickResign();
		defaultFederate.quickResign();
		defaultFederate.quickDestroy();
	}
	
	@AfterClass(alwaysRun=true)
	public void afterClass()
	{
		super.afterClass();
	}
	
	///////////////////////////////////////////////////////////////////////////////////////////
	///////////////////////////////// DDM Suport Test Methods /////////////////////////////////
	///////////////////////////////////////////////////////////////////////////////////////////
	// public int getRoutingSpaceHandle( String theName )
	//        throws NameNotFound,
	//               FederateNotExecutionMember,
	//               RTIinternalError;
	@Test
	public void testGetRoutingSpaceHandle() throws Exception
	{
		Assert.assertEquals( defaultFederate.rtiamb.getRoutingSpaceHandle("TestSpace"), spaceHandle,
		                     "Received incorrect routing space handle in default federate" );
		Assert.assertEquals( secondFederate.rtiamb.getRoutingSpaceHandle("TestSpace"), spaceHandle,
		                     "Received incorrect routing space handle in second federate" );
	}

	@Test
	public void testGetRoutingSpaceHandleWhenNotJoined()
	{
		secondFederate.quickResign();
		try
		{
			secondFederate.rtiamb.getRoutingSpaceHandle( "TestSpace" );
			expectedException( FederateNotExecutionMember.class );
		}
		catch( FederateNotExecutionMember fnem )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, FederateNotExecutionMember.class );
		}
	}

	@Test
	public void testGetRoutingSpaceHandleWithNonExistentName()
	{
		try
		{
			defaultFederate.rtiamb.getRoutingSpaceHandle( "NoSuchSpace" );
			expectedException( NameNotFound.class );
		}
		catch( NameNotFound nnf )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, NameNotFound.class );
		}
	}

	@Test
	public void testGetRoutingSpaceHandleWithNullName()
	{
		try
		{
			defaultFederate.rtiamb.getRoutingSpaceHandle( null );
			expectedException( NameNotFound.class );
		}
		catch( NameNotFound nnf )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, NameNotFound.class );
		}
	}

	// public String getRoutingSpaceName( int theHandle )
	//        throws SpaceNotDefined,
	//               FederateNotExecutionMember,
	//               RTIinternalError;
	@Test
	public void testGetRoutingSpaceName() throws Exception
	{
		Assert.assertEquals( defaultFederate.rtiamb.getRoutingSpaceName(spaceHandle), "TestSpace",
		                     "Incorrect routing space name received from RTI by defaultFederate" );
		Assert.assertEquals( secondFederate.rtiamb.getRoutingSpaceName(spaceHandle), "TestSpace",
		                     "Incorrect routing space name received from RTI by secondFederate" );
	}

	@Test
	public void testGetRoutingSpaceNameWhenNotJoined()
	{
		secondFederate.quickResign();
		try
		{
			secondFederate.rtiamb.getRoutingSpaceName( spaceHandle );
			expectedException( FederateNotExecutionMember.class );
		}
		catch( FederateNotExecutionMember fnem )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, FederateNotExecutionMember.class );
		}
	}

	@Test
	public void testGetRoutingSpaceNameWithNonExistentHandle()
	{
		try
		{
			defaultFederate.rtiamb.getRoutingSpaceName( 1000000 );
			expectedException( SpaceNotDefined.class );
		}
		catch( SpaceNotDefined snd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, SpaceNotDefined.class );
		}
	}
	
	@Test
	public void testGetRoutingSpaceNameWithNegativeHandle()
	{
		try
		{
			defaultFederate.rtiamb.getRoutingSpaceName( -1 );
			expectedException( SpaceNotDefined.class );
		}
		catch( SpaceNotDefined snd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, SpaceNotDefined.class );
		}
	}

	// public int getDimensionHandle( String theName, int whichSpace )
	//        throws SpaceNotDefined,
	//               NameNotFound,
	//               FederateNotExecutionMember,
	//               RTIinternalError;
	@Test
	public void testGetDimensionHandle() throws Exception
	{
		Assert.assertEquals( defaultFederate.rtiamb.getDimensionHandle("TestDimension",spaceHandle),
		                     dimensionHandle,
		                     "Wrong dimension handle received from RTI in defaultFederate" );

		Assert.assertEquals( secondFederate.rtiamb.getDimensionHandle("TestDimension",spaceHandle),
		                     dimensionHandle,
		                     "Wrong dimension handle received from RTI in secondFederate" );

	}

	@Test
	public void testGetDimensionHandleWhenNotJoined()
	{
		secondFederate.quickResign();
		try
		{
			secondFederate.rtiamb.getDimensionHandle( "TestDimension", spaceHandle );
			expectedException( FederateNotExecutionMember.class );
		}
		catch( FederateNotExecutionMember fnem )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, FederateNotExecutionMember.class );
		}
	}

	@Test
	public void testGetDimensionHandleWithNonExistentName()
	{
		try
		{
			defaultFederate.rtiamb.getDimensionHandle( "NoSuchDimension", spaceHandle );
			expectedException( NameNotFound.class );
		}
		catch( NameNotFound nnf )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, NameNotFound.class );
		}
	}
	
	@Test
	public void testGetDimensionHandleWithNullName()
	{
		try
		{
			defaultFederate.rtiamb.getDimensionHandle( null, spaceHandle );
			expectedException( NameNotFound.class );
		}
		catch( NameNotFound nnf )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, NameNotFound.class );
		}
	}
	
	@Test
	public void testGetDimensionHandleWithNonExistentSpace()
	{
		try
		{
			defaultFederate.rtiamb.getDimensionHandle( "TestDimension", 1000000 );
			expectedException( SpaceNotDefined.class );
		}
		catch( SpaceNotDefined snd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, SpaceNotDefined.class );
		}
	}
	
	@Test
	public void testGetDimensionHandleWithNegativeSpace()
	{
		try
		{
			defaultFederate.rtiamb.getDimensionHandle( "TestDimension", -1 );
			expectedException( SpaceNotDefined.class );
		}
		catch( SpaceNotDefined snd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, SpaceNotDefined.class );
		}
	}

	// public String getDimensionName( int theHandle, int whichClass )
	//        throws SpaceNotDefined,
	//               DimensionNotDefined,
	//               FederateNotExecutionMember,
	//               RTIinternalError;
	@Test
	public void testGetDimensionName() throws Exception
	{
		Assert.assertEquals( defaultFederate.rtiamb.getDimensionName(dimensionHandle,spaceHandle),
		                     "TestDimension",
		                     "Wrong dimension name received from RTI in defaultFederate" );

		Assert.assertEquals( secondFederate.rtiamb.getDimensionName(dimensionHandle,spaceHandle),
		                     "TestDimension",
		                     "Wrong dimension name received from RTI in secondFederate" );
	}

	@Test
	public void testGetDimensionNameWhenNotJoined()
	{
		secondFederate.quickResign();
		try
		{
			secondFederate.rtiamb.getDimensionName( dimensionHandle, spaceHandle );
			expectedException( FederateNotExecutionMember.class );
		}
		catch( FederateNotExecutionMember fnem )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, FederateNotExecutionMember.class );
		}
	}

	@Test
	public void testGetDimensionNameWithNonExistentDimension()
	{
		try
		{
			defaultFederate.rtiamb.getDimensionName( 1000000, spaceHandle );
			expectedException( DimensionNotDefined.class );
		}
		catch( DimensionNotDefined dnd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, DimensionNotDefined.class );
		}
	}
	
	@Test
	public void testGetDimensionNameWithNegativeDimension()
	{
		try
		{
			defaultFederate.rtiamb.getDimensionName( -1, spaceHandle );
			expectedException( DimensionNotDefined.class );
		}
		catch( DimensionNotDefined dnd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, DimensionNotDefined.class );
		}
	}
	
	@Test
	public void testGetDimensionNameWithNonExistentSpace()
	{
		try
		{
			defaultFederate.rtiamb.getDimensionName( dimensionHandle, 100000 );
			expectedException( SpaceNotDefined.class );
		}
		catch( SpaceNotDefined snd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, SpaceNotDefined.class );
		}
	}
	
	@Test
	public void testGetDimensionNameWithNegativeSpace()
	{
		try
		{
			defaultFederate.rtiamb.getDimensionName( dimensionHandle, -1 );
			expectedException( SpaceNotDefined.class );
		}
		catch( SpaceNotDefined snd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, SpaceNotDefined.class );
		}
	}

	// public int getAttributeRoutingSpaceHandle( int theHandle, int whichClass )
	//        throws ObjectClassNotDefined,
	//               AttributeNotDefined,
	//               FederateNotExecutionMember,
	//               RTIinternalError;

	// not yet implemented

	// public int getInteractionRoutingSpaceHandle( int theHandle )
	//        throws InteractionClassNotDefined,
	//               FederateNotExecutionMember,
	//               RTIinternalError;

	// not yet implemented

	// public Region getRegion( int regionToken )
	//        throws FederateNotExecutionMember,
	//               ConcurrentAccessAttempted,
	//               RegionNotKnown,
	//               RTIinternalError;
	@Test
	public void testGetRegion() throws Exception
	{
		Assert.assertEquals( defaultFederate.rtiamb.getRegion(regionHandle), region,
		                     "Wrong Region reference received from the RTI in defaultFederate" );
	}

	@Test
	public void testGetRegionWithUnknownHandleInFederate()
	{
		// now try with the second federate. should result in an exception
		try
		{
			secondFederate.rtiamb.getRegion( regionHandle );
			expectedException( RegionNotKnown.class );
		}
		catch( RegionNotKnown rnn )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, RegionNotKnown.class );
		}
	}
	
	@Test
	public void testGetRegionWhenNotJoined()
	{
		defaultFederate.quickResign();
		try
		{
			defaultFederate.rtiamb.getRegion( regionHandle );
			expectedException( FederateNotExecutionMember.class );
		}
		catch( FederateNotExecutionMember fnem )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, FederateNotExecutionMember.class );
		}
	}

	@Test
	public void testGetRegionWithNonExistentRegionToken()
	{
		try
		{
			secondFederate.rtiamb.getRegion( 100000 );
			expectedException( RegionNotKnown.class );
		}
		catch( RegionNotKnown rnn )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, RegionNotKnown.class );
		}
	}
	
	@Test
	public void testGetRegionWithNegativeRegionToken()
	{
		try
		{
			secondFederate.rtiamb.getRegion( -1 );
			expectedException( RegionNotKnown.class );
		}
		catch( RegionNotKnown rnn )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, RegionNotKnown.class );
		}
	}

	// public int getRegionToken( Region region )
	//        throws FederateNotExecutionMember,
	//               ConcurrentAccessAttempted,
	//               RegionNotKnown,
	//               RTIinternalError;
	@Test
	public void testGetRegionToken() throws Exception
	{
		Assert.assertEquals( defaultFederate.rtiamb.getRegionToken(region), regionHandle,
		                     "Wrong region token received from RTI in defaultFederate" );
	}

	@Test
	public void testGetRegionTokenInvalidInFederate()
	{
		try
		{
			secondFederate.rtiamb.getRegionToken( region );
			expectedException( RegionNotKnown.class );
		}
		catch( RegionNotKnown rnn )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, RegionNotKnown.class );
		}
	}
	
	@Test
	public void testGetRegionTokenWhenNotJoined()
	{
		defaultFederate.quickResign();
		try
		{
			defaultFederate.rtiamb.getRegionToken( region );
			expectedException( FederateNotExecutionMember.class );
		}
		catch( FederateNotExecutionMember fnem )
		{
			// succcess!
		}
		catch( Exception e )
		{
			wrongException( e, FederateNotExecutionMember.class );
		}
	}

	@Test
	public void testGetRegionTokenWithNullRegion()
	{
		try
		{
			defaultFederate.rtiamb.getRegionToken( null );
			expectedException( RegionNotKnown.class );
		}
		catch( RegionNotKnown rnn )
		{
			// succcess!
		}
		catch( Exception e )
		{
			wrongException( e, RegionNotKnown.class );
		}
	}

	//----------------------------------------------------------
	//                     STATIC METHODS
	//----------------------------------------------------------
}
